utils/monitoring.js

```diff
@@ -0,0 +1,88 @@
+const { errors } = require('arsenal');
+const promClient = require('prom-client');
+const { Registry } = require('prom-client');
+const { http } = require('httpagent');
+const CountMaster = require('../CountItems/CountMaster');
+
+const aggregatorRegistry = new promClient.AggregatorRegistry();
+const { collectDefaultMetrics } = promClient;
+
+
+// Histogram of the bucket processing duration, by the utilization service.
+const bucketProcessingDuration = new promClient.Histogram({
+    name: 's3_countitems_bucket_listing_duration_seconds',
+    help: 'Bucket processing duration',
+    buckets: [1, 10, 60, 600, 3600, 18000, 36000],
+});
+
+const consolidationDuration = new promClient.Histogram({
+    name: 's3_countitems_bucket_merge_duration_seconds',
+    help: 'Duration of metrics consolidation in seconds',
+    buckets: [0.01, 0.05, 0.1, 0.2, 0.5, 1, 2, 5, 10],
+});
+
+const workersCount = new promClient.Counter({
+    name: 's3_countitems_total_buckets_count',
+    help: 'Buckets processed by workers state',
+    labelNames: ['state'],
+});
+
+const objectsCount = new promClient.Counter({
+    name: 's3_countitems_total_objects_count',
+    help: 'Number of objects processed',
+    labelNames: ['status'],
+});
+
+/**
+ * @param {http.ServerResponse} res - http response object
+ * @param {Error | errors.ArsenalError} error - Error
+ * @return {void}
+ */
+function _writeResponse(res, error) {
+    let statusCode = 500;
+    if (error instanceof errors.ArsenalError && Number.isInteger(error.code)) {
+        statusCode = error.code;
+    }
+    res.writeHead(statusCode, { 'Content-Type': 'application/json' });
+    res.end();
+}
+
+/**
+ * @param {function} onScraped - callback to call when metrics are scraped
+ * @param {http.IncomingMessage} req - http request object
+ * @param {http.ServerResponse} res - http response object
+ * @return {void}
+ */
+async function metricsHandler(onScraped, req, res) {
+    if (req.method !== 'GET' || req.url !== '/metrics') {
+        return _writeResponse(res, errors.MethodNotAllowed);
+    }
+    try {
+        const [registerMetrics, clusterMetrics] = await Promise.all([
+            promClient.register.metrics(),
+            aggregatorRegistry.clusterMetrics(),
+        ]);
+        const promMetrics = `${registerMetrics}\n${clusterMetrics}`;
+        const contentLen = Buffer.byteLength(promMetrics, 'utf8');
+        return res.writeHead(200, {
+            'content-length': contentLen,
+            'content-type': promClient.register.contentType,
+        }).end(promMetrics);
+    } catch (ex) {
+        return _writeResponse(res, ex);
+    } finally {
+        if (onScraped) {
+            onScraped();
+        }
```

Comment:
```suggestion
        return onScraped?.();
```

Maybe we also want to make sure it's a function

Reply:
why would it not be a function: if it's provided but not a function, this is out of the contract?
(similar to every other callback : we almost never check if they are indeed callbacks...)

Reply:
It's more to be on the safe side: we are not using typescript, the argument seems to be optional, but first in the list, as we have a condition on its value... Given that it's called outside of the try/catch, it would throw an error to call a function if the object is not one, independently from any contract we might have (maybe we can reuse the function in other s3utils scripts...). And we already had bugs where callbacks were "not a function" in some cases...

Reply:
that is my point: i don't think it helps...
* if somebody passes a value instead of the callback, he expects something with it (either he though this was a callback, or he wanted to pass extra info): so silently ignoring the unexpected type will just bury the bug...
* if we throw an error, the process will crash (or at the very least display a log about the uncaught exception) which will be easier to notice while testing
* obviously this if the code is not tested at all (either manually or automated), then all odds are off...

yes typescript will be better ("mandatory" compile stage would immediately detect the mismatch), but since we don't have it we must be more thorough when _testing_ instead of trying to handle unrecoverable errors...

Reply:
To me, the function should not be dependent on the whole project implementation:
* First, because we don't test a lot s3utils
* Second, because s3utils acts as a repo of multiple independent projects: we don't want any kind of contract with the rest of the implementation, we just want to make sure the object is a function before calling it...
* And third, because our functions should be robust, introducing such contract is only known by the ones implementing the function... In JS we can have plenty of reason where an object can be undefined, and we don't want to throw because of that. If we still want to log something, we can log it in this case using  `error` level.

Anyway it's not blocking, as I approved the PR